### PR TITLE
added documentation for public parts of the crate

### PR DESCRIPTION
**Pull request by serejkus**
Added docs for public parts of the crate, as requested in #17 and #23.

English is not my native language, so feel free to fix grammar issues. The same applies to examples: most of them are just copy-paste and involve parsing of semver structs from strings, maybe there might be a better approach in some cases.

### CODE REVIEW

**Comment by serejkus** [comment]
@KodrAus, any suggestions on this PR?

**Comment by KodrAus** [comment]
Sorry I somehow missed this. Thanks for the ping @serejkus and for taking on these docs! I'll dig in to this later today.

File: src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,52 @@
+//! Collection of structures and helper functions for parsing semantic version.
+//!
+//! This crate contains data structures for holding version data and comparision of versions
+//! according to [Semantic versioning](http://semver.org/).
+//! Main structs are [`version::Version`] representing version itself and [`range::VersionReq`] as a
+//! collection of [`range::Predicate`]
+//! representing data for version comparision.
+//!
+//! # Example
```
Comment (KodrAus) [inline src/lib.rs]: Let's change this to _Examples_ to be consistent.

File: src/lib.rs [hunk 2]
```diff
@@ -1,3 +1,52 @@
+//! Collection of structures and helper functions for parsing semantic version.
+//!
+//! This crate contains data structures for holding version data and comparision of versions
```
Comment (KodrAus) [inline src/lib.rs]: Typo here: _comparision_ to _comparison_. I always type comparision for some reason...

Looks like there are a few other places we should replace _comparision_ with _comparison_.

File: src/lib.rs [hunk 3]
```diff
@@ -1,3 +1,52 @@
...
+//! representing data for version comparision.
+//!
+//! # Example
+//!
+//! ```rust
+//! use semver_parser::range;
+//! use semver_parser::version;
+//!
+//! # fn try_main() -> Result<(), String> {
+//!
```
Comment (KodrAus) [inline src/lib.rs]: We should remove these newlines after hidden lines, it makes the code here look a little funny, but when it renders in the docs you only have 1 visible newline.

Does that make sense?

File: src/lib.rs [hunk 4]
```diff
@@ -1,3 +1,52 @@
...
+//! This crate contains data structures for holding version data and comparision of versions
+//! according to [Semantic versioning](http://semver.org/).
+//! Main structs are [`version::Version`] representing version itself and [`range::VersionReq`] as a
+//! collection of [`range::Predicate`]
+//! representing data for version comparision.
+//!
+//! # Example
+//!
+//! ```rust
```
Comment (KodrAus) [inline src/lib.rs]: For the docs we don't actually need the `rust` annotation here. `rustdoc` assumes the code block contains testable Rust by default.

File: src/range.rs [hunk 5]
```diff
@@ -1,30 +1,222 @@
+//! Version range and requirements data and functions (used for version comparision)
```
Comment (KodrAus) [inline src/range.rs]: Let's throw a `.` at the end of these introduction lines in all the modules, structures and methods to be consistent.

File: src/version.rs [hunk 6]
```diff
@@ -1,19 +1,108 @@
...
+/// # Ok(())
+/// # }
+/// #
+/// # fn main() {
+/// #   try_main().unwrap();
+/// # }
+/// ```
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct Version {
+    /// major version as number (`0` in `"0.1.2"`);
```
Comment (KodrAus) [inline src/version.rs]: Let's capitalise the _major_, _minor_ etc on these docs and use a `.` instead of a `;` at the end.

**Review by KodrAus — COMMENTED** [review]
Thanks @serejkus this is an awesome effort! I've just left some comments for a coupe of typos.

**Comment by serejkus** [comment]
Thanks for pointing out! Will try to fix those issues in a day.

File: src/lib.rs (hunk 1 above)
Comment (steveklabnik) [reply src/lib.rs]: In general, all of them should always be `Examples` :)

**Comment by steveklabnik** [comment]
Seconded, thanks a ton @serejkus :)

**Comment by serejkus** [comment]
Fixed issues with style and added links to cargo docs explaining caret and tilde version requirements. Two commits are ok or is it better to rebase it into a single commit?

**Comment by steveklabnik** [comment]
Two commits is just fine. Thanks again!

**Pull request merged**